Add n.exchange centralized swap integration

MMrj9 opened the pull request:
# Add n.exchange centralized swap integration

## Description

Integrate n.exchange as a centralized swap provider supporting multiple networks and tokens. The implementation uses contract addresses for token identification, eliminating the need for currency code mapping. This approach provides better support for tokens and aligns with API requirements.

The integration includes support for native currencies and ERC20 tokens across major networks including Ethereum, Polygon, Base, Arbitrum, Optimism, BSC, and others. The plugin handles rate queries, order creation, and payment processing through the n.exchange API v2.

Additional changes include mapctl tooling for automatic provider mappings and updates to existing swap plugins to use the new mapping system.

## CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

## Dependencies

none

## Changes

- Add n.exchange centralized swap plugin implementation
- Support for contract address-based token identification
- Integration with n.exchange API v2 for rate queries and order management
- Support for multiple networks: Ethereum, Polygon, Base, Arbitrum, Optimism, BSC, and others
- Mapctl tooling for automatic provider mappings
- Updates to existing swap plugins to use new mapping system

## Testing

- [x] Unit tests added/updated (nexchange tests included in partnerJson.test.ts)
- [ ] Manual testing completed

## Checklist

- [x] Code follows project style guidelines
- [x] Commit message follows git commit guidelines
- [x] Branch is synced with upstream master
- [x] Tests pass locally (26 passing)
- [x] Documentation updated (CHANGELOG.md updated)
- [x] Linting passes
- [x] Type checking passes (nexchange-specific types fixed)


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Introduces a new external swap integration that constructs spend targets and order metadata based on third-party API responses, so parsing/mapping errors could impact swap availability and correctness. Changes are additive and isolated to a new plugin plus wiring/tests.
> 
> **Overview**
> Adds **n.exchange** as a new centralized swap provider (`nexchange`) and wires it into `src/index.ts` so it is available alongside existing swap plugins.
> 
> Implements a full API v2 integration that fetches fixed-rate quotes, enforces min/max limits, creates orders, and builds `EdgeSpendInfo`/swap metadata, using **contract-address-based token identification** (plus a pluginId→network-code transcription map) for native assets and tokens.
> 
> Updates test/config coverage by adding a `nexchange` partner mapping JSON, extending `partnerJson.test.ts` with nexchange cases, and adding `NEXCHANGE_INIT` options for API key/referral configuration; also documents the addition in `CHANGELOG.md`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3148f316f64577be795a65694caac19e2af6a80a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

MMrj9 commented on src/swap/central/nexchange.ts:
```diff
@@ -0,0 +1,553 @@
+    async fetchSwapQuote(
+      req: EdgeSwapRequest,
+      userSettings: Object | undefined,
+      opts: { promoCode?: string }
```
promoCode support can be added at some point

paullinator reviewed (changes requested):
n/a

paullinator commented on src/swap/central/nexchange.ts:
```diff
@@ -0,0 +1,553 @@
+    // Check if rate is expired
+    const expirationTime = parseInt(rate.expiration_time_unix, 10) * 1000
+    if (Date.now() >= expirationTime) {
+      throw new SwapCurrencyError(swapInfo, request)
```
**Critical: parseInt Could Return NaN**

If `rate.expiration_time_unix` is an empty string or non-numeric, `parseInt` returns `NaN`, and `Date.now() >= NaN` is always `false`, so expired rates pass validation.

**Recommendation:**
```typescript
const expirationTime = parseInt(rate.expiration_time_unix, 10) * 1000
if (Number.isNaN(expirationTime) || Date.now() >= expirationTime) {
  throw new SwapCurrencyError(swapInfo, request)
}
```

MMrj9 replied:
has been applied

paullinator commented on src/swap/central/nexchange.ts:
```diff
@@ -0,0 +1,553 @@
+  max_deposit_amount: asString,
+  min_deposit_amount: asString,
+  expiration_time_unix: asString
+})
```
**Suggestion: Unused Fields in Cleaners**

The `asRateV2` cleaner has unused fields: `pair`, `from`, `to`, `withdrawal_fee`. Same issue with `asOrderV2` (lines 195-223) which has many unused fields.

**Recommendation:** Remove unused fields or add a comment explaining they're validated for future use.

MMrj9 replied:
added comment

paullinator commented on src/swap/central/nexchange.ts:
```diff
@@ -0,0 +1,553 @@
+      )
+    ])
+
+    // Get network codes from plugin IDs (no currency codes needed - using contract addresses only)
```
**Suggestion: Misleading Comment**

The comment says "using contract addresses only" but the code actually uses empty strings for native currencies.

**Recommendation:** Change to:
```typescript
// Get network codes from plugin IDs (contract addresses used for tokens, network codes for native currencies)
```

paullinator reviewed (changes requested):
## Review Summary

The implementation is generally well-structured and follows existing plugin patterns. However, there are **3 critical issues** that should be addressed before merge.

### Critical Issues (Must Fix)
1. JSON parsing errors not handled (line 269)
2. `getContractAddress` silently falls back to native currency (lines 141-149)
3. `parseInt` could return NaN, bypassing expiration check (lines 349-351)

### Additional Note
**Commit Message Inaccuracy:** The commit body mentions "updates to existing swap plugins to use the new mapping system" but there are no modifications to existing swap plugins in this commit. Consider amending the commit message for accuracy.

### Positive Observations
- Good use of existing helpers (`getMaxSwappable`, `makeSwapPluginQuote`, etc.)
- Clean cleaner definitions with `asRateV2` and `asOrderV2`
- Proper error handling with `SwapCurrencyError`, `SwapAboveLimitError`, `SwapBelowLimitError`
- Good use of `Promise.all` for parallel address fetching
- Appropriate use of `ensureInFuture` for expiration handling
- Good memo handling for chains like XRP/XLM

See inline comments for specific issues and recommendations.

paullinator reviewed (commented):
## Review Summary

This PR adds n.exchange as a new centralized swap provider. The implementation uses contract addresses for token identification and integrates with the n.exchange API v2.

### Previous Review Comments - All Addressed ✅

All critical and warning issues raised by Paul and Bug Bot have been addressed:
- `getContractAddress` now throws errors instead of silently falling back
- `parseInt` NaN check added
- JSON parsing wrapped in try/catch
- Cleaner failures handled with try/catch
- `checkInvalidTokenIds` called
- `ensureInFuture` used for expiration dates

### Remaining Items

**Before merge:**
1. Add `checkWhitelistedMainnetCodes` call for consistency with other plugins
2. Consider moving `MAINNET_CODE_TRANSCRIPTION` to `src/mappings/nexchange.ts`

**Nice to have (can be deferred):**
- Add unit tests for `getFixedQuote()` business logic
- Consolidate repetitive documentation comments

### Git

⚠️ The PR contains a merge commit instead of a rebase. Consider rebasing onto master before merge:
```bash
git fetch origin master
git rebase origin/master
git push --force-with-lease
```

**Ready for merge** after adding `checkWhitelistedMainnetCodes`.

---

See inline comments for specific suggestions.

paullinator commented on src/swap/central/nexchange.ts:
```diff
@@ -0,0 +1,573 @@
+// See https://api.n.exchange/en/api/v2/currency/ for list of supported currencies
+// Network codes map to Nexchange network identifiers
+// Based on supported networks: ALGO, ATOM, SOL, BCH, BTC, DASH, DOGE, DOT, EOS, TON, HBAR, LTC, XLM, XRP, XTZ, ZEC, TRON, ADA, BASE, MATIC/POL, ETH, AVAXC, BSC, ETC, ARB, OP, FTM, SONIC
+// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
```
**Suggestion:** Other plugins use mapping files in `src/mappings/` with `mapToRecord()`. Consider creating `src/mappings/nexchange.ts` for consistency:

```ts
import { mapToRecord } from '../utils'

export const MAINNET_CODE_TRANSCRIPTION = mapToRecord([
  ['ethereum', 'ETH'],
  // ...
])
```

paullinator commented on src/swap/central/nexchange.ts:
```diff
@@ -0,0 +1,573 @@
+    } catch (e) {
+      throw new Error(`Nexchange rate response parsing failed: ${String(e)}`)
+    }
+    const rate = rates[0] // Contract address query returns single rate
```
**Suggestion:** This assumes the API returns exactly one rate. Consider validating array length:

```ts
if (rates.length === 0) {
  throw new SwapCurrencyError(swapInfo, request)
}
const rate = rates[0]
```

paullinator reviewed (commented):
## Additional Inline Suggestions

See inline comments for specific code improvements. These are minor suggestions - the main review summary was posted earlier.

**Key inline items:**
- Line 560: Add `checkWhitelistedMainnetCodes` call for consistency
- Line 80: Consider using mapping file pattern
- Various type safety and naming improvements

paullinator reviewed (changes requested):
see comments

MMrj9 committed:
Add n.exchange centralized swap integration

Integrate n.exchange as a centralized swap provider supporting multiple
networks and tokens. The implementation uses contract addresses for token
identification, eliminating the need for currency code mapping.

Changes:
- Add n.exchange swap plugin (src/swap/central/nexchange.ts)
- Add n.exchange mapping file (src/mappings/nexchange.ts)
- Register nexchange plugin in src/index.ts
- Add partner mapping JSON and tests
- Update CHANGELOG.md

samholmes commented:
This work was completed by https://github.com/EdgeApp/edge-exchange-plugins/pull/437 and is released in version 2.42.0 of edge-exchange-plugins
